indico/modules/rb/client/js/common/rooms/RoomEditModal.jsx

```diff
@@ -8,951 +8,200 @@
+      },
+      {
+        menuItem: 'Location',
+        pane: {key: 'location', content: <RoomEditLocation />},
+      },
+      {
+        menuItem: 'Permissions',
+        pane: {key: 'permissions', content: <RoomEditPermissions />},
+      },
+      {
+        key: 'notifications',
+        menuItem: 'Notifications',
+        pane: {
+          key: 'notifications',
+          content: <RoomEditNotifications />,
+        },
+      },
+      {
+        key: 'options',
+        menuItem: 'Options',
+        pane: {
+          key: 'options',
+          content: <RoomEditOptions showEquipment globalAttributes={globalAttributes} />,
+        },
+      },
+    ],
+    [globalAttributes]
+  );
 
-  renderModalContent = fprops => {
-    const {afterCreation} = this.props;
-    const {hasValidationErrors, pristine, submitting} = fprops;
-    const {submitState, wasEverDirty} = this.state;
+  const renderModal = formProps => {
+    const {
+      handleSubmit: handleFormSubmit,
+      hasValidationErrors,
+      pristine,
+      submitting,
+      dirty,
+    } = formProps;
     return (
-      <>
+      <Modal open onClose={closeModal} size="large" centered={false} closeIcon>
         <Modal.Header>
-          {this.newRoom ? Translate.string('Add Room') : Translate.string('Edit Room Details')}
+          {isNewRoom() ? <Translate>Add Room</Translate> : <Translate>Edit Room Details</Translate>}
         </Modal.Header>
-        <Modal.Content scrolling>
-          <Form id="room-form" styleName="room-form" onSubmit={fprops.handleSubmit}>
-            <Grid columns={3}>
-              <Grid.Column>{columns[0].map(this.renderContent)}</Grid.Column>
-              <Grid.Column>{columns[1].map(this.renderContent)}</Grid.Column>
-              <Grid.Column>
-                {columns[2].map(this.renderContent)}
-                <Message
-                  styleName="submit-message"
-                  positive
-                  hidden={!afterCreation || wasEverDirty}
-                >
-                  <Translate>Room has been successfully created.</Translate>
-                </Message>
-                <Message styleName="submit-message" positive hidden={submitState !== 'success'}>
+        <Modal.Content>
+          <Message styleName="submit-message" positive hidden={!afterCreation || wasEverUpdated}>
+            <Translate>Room has been successfully created.</Translate>
+          </Message>
+          <FormSpy subscription={{submitSucceeded: true}}>
+            {({submitSucceeded}) => {
+              return (
+                <Message styleName="submit-message" positive hidden={!submitSucceeded || dirty}>
                   <Translate>Room has been successfully updated.</Translate>
                 </Message>
-                <Message styleName="submit-message" negative hidden={submitState !== 'error'}>
+              );
+            }}
+          </FormSpy>
+          <FormSpy subscription={{submitFailed: true}}>
```

Comment:
I'd use a single FormSpy that subscribes to both `submitSucceeded` and `submitFailed`.

In principle you can probably avoid using `FormSpy` altogether and just add it to the global subscription in the FinalForm and then get these two values from the fprops.

Reply:
I've had both approaches, ended up with this because using two `FormSpy`'s will result in an extra empty `<>` to hold both `Message` components (unless we merge them and use conditionals).
Adding it as a global subscription would be perfect, but wouldn't that make the whole form re-render when we only need it in a small part?